GenericContainer#withCopyFileToContainer is not efficient memory-wise

Hi,

We've recently seen OOM in our tests and after debugging  by generating a heap dump and opening it, we found that for the following code that we use in XWiki tests, TC is copying all the content of the `sourceDirectory` in memory:

```
MountableFile mountableDirectory = MountableFile.forHostPath(sourceDirectory);
container.withCopyFileToContainer(mountableDirectory, targetDirectory);
```

Since our directory contains 300MB+ of files, the memory we give to the JVM is not enough.

Note that we copy and don't bind because we need to have it work in the DOOD use case. Our code is:

```
    protected void mountFromHostToContainer(GenericContainer<?> container, String sourceDirectory,
        String targetDirectory)
    {
        // Note 1: File mounting is awfully slow on Mac OSX. For example starting Tomcat with XWiki mounted takes
        // 45s+, while doing a COPY first and then starting Tomcat takes 8s (+5s for the copy).
        // Note 2: For the DOOD use case, we also do the copy instead of the volume mounting since that would require
        // to have the sourceDirectory path mounted from the host and this would put and leave files on the host which
        // would not work with parallel executions (think about multiple CI jobs executing in parallel on the same host)
        // and would also not be clean.
        String osName = System.getProperty("os.name").toLowerCase();
        if (isInAContainer() || osName.startsWith("mac os x")) {
            MountableFile mountableDirectory = MountableFile.forHostPath(sourceDirectory);
            container.withCopyFileToContainer(mountableDirectory, targetDirectory);
        } else {
            container.withFileSystemBind(sourceDirectory, targetDirectory);
        }
    }
```

It would be great if TC could stream the data instead of keeping it in memory. Is that possible?

Thanks

Hello guys. FTR we're still getting this issue on the XWiki build.

```
01:36:19,047 [ERROR] There was an error in the forked process
01:36:19,047 [ERROR] Java heap space
01:36:19,047 [ERROR] java.lang.OutOfMemoryError: Java heap space
01:36:19,047 [ERROR]    at java.base/java.util.Arrays.copyOf(Arrays.java:3745)
01:36:19,047 [ERROR]    at java.base/java.io.ByteArrayOutputStream.grow(ByteArrayOutputStream.java:120)
01:36:19,047 [ERROR]    at java.base/java.io.ByteArrayOutputStream.ensureCapacity(ByteArrayOutputStream.java:95)
01:36:19,047 [ERROR]    at java.base/java.io.ByteArrayOutputStream.write(ByteArrayOutputStream.java:156)
01:36:19,047 [ERROR]    at org.apache.commons.compress.utils.CountingOutputStream.write(CountingOutputStream.java:62)
01:36:19,047 [ERROR]    at org.apache.commons.compress.utils.FixedLengthBlockOutputStream$BufferAtATimeOutputChannel.write(FixedLengthBlockOutputStream.java:91)
01:36:19,047 [ERROR]    at org.apache.commons.compress.utils.FixedLengthBlockOutputStream.writeBlock(FixedLengthBlockOutputStream.java:259)
01:36:19,047 [ERROR]    at org.apache.commons.compress.utils.FixedLengthBlockOutputStream.maybeFlush(FixedLengthBlockOutputStream.java:169)
01:36:19,047 [ERROR]    at org.apache.commons.compress.utils.FixedLengthBlockOutputStream.write(FixedLengthBlockOutputStream.java:206)
01:36:19,047 [ERROR]    at org.apache.commons.compress.archivers.tar.TarArchiveOutputStream.write(TarArchiveOutputStream.java:713)
01:36:19,047 [ERROR]    at java.base/java.io.InputStream.transferTo(InputStream.java:705)
01:36:19,047 [ERROR]    at java.base/java.nio.file.Files.copy(Files.java:3119)
01:36:19,047 [ERROR]    at org.testcontainers.utility.MountableFile.recursiveTar(MountableFile.java:362)
01:36:19,047 [ERROR]    at org.testcontainers.utility.MountableFile.recursiveTar(MountableFile.java:371)
01:36:19,047 [ERROR]    at org.testcontainers.utility.MountableFile.recursiveTar(MountableFile.java:371)
01:36:19,047 [ERROR]    at org.testcontainers.utility.MountableFile.recursiveTar(MountableFile.java:371)
01:36:19,047 [ERROR]    at org.testcontainers.utility.MountableFile.transferTo(MountableFile.java:327)
01:36:19,047 [ERROR]    at org.testcontainers.containers.ContainerState.copyFileToContainer(ContainerState.java:306)
01:36:19,047 [ERROR]    at org.testcontainers.containers.ContainerState.copyFileToContainer(ContainerState.java:282)
01:36:19,047 [ERROR]    at org.testcontainers.containers.GenericContainer$$Lambda$603/0x00000001003ca040.accept(Unknown Source)
01:36:19,047 [ERROR]    at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
01:36:19,047 [ERROR]    at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:430)
01:36:19,047 [ERROR]    at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:344)
01:36:19,047 [ERROR]    at org.testcontainers.containers.GenericContainer$$Lambda$571/0x00000001003a1440.call(Unknown Source)
01:36:19,047 [ERROR]    at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
01:36:19,047 [ERROR]    at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:334)
01:36:19,047 [ERROR]    at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:322)
01:36:19,047 [ERROR]    at org.xwiki.test.docker.internal.junit5.DockerTestUtils.startContainerInternal(DockerTestUtils.java:218)
01:36:19,047 [ERROR]    at org.xwiki.test.docker.internal.junit5.DockerTestUtils.startContainer(DockerTestUtils.java:181)
01:36:19,047 [ERROR]    at org.xwiki.test.docker.internal.junit5.AbstractContainerExecutor.start(AbstractContainerExecutor.java:56)
01:36:19,047 [ERROR]    at org.xwiki.test.docker.internal.junit5.servletengine.ServletContainerExecutor.startContainer(ServletContainerExecutor.java:323)
01:36:19,048 [ERROR]    at org.xwiki.test.docker.internal.junit5.servletengine.ServletContainerExecutor.start(ServletContainerExecutor.java:155)
01:36:19,048 [ERROR]
01:36:19,048 [ERROR]    at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:628)
01:36:19,048 [ERROR]    at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:285)
01:36:19,048 [ERROR]    at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:250)
01:36:19,048 [ERROR]    at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1203)
01:36:19,048 [ERROR]    at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1055)
01:36:19,048 [ERROR]    at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:871)
```

--- a/core/src/main/java/org/testcontainers/containers/ContainerState.java
+++ b/core/src/main/java/org/testcontainers/containers/ContainerState.java
@@ -22,17 +22,14 @@ import org.testcontainers.utility.LogUtils;
 import org.testcontainers.utility.MountableFile;
 import org.testcontainers.utility.ThrowingFunction;
 
-import java.io.BufferedInputStream;
-import java.io.BufferedOutputStream;
 import java.io.File;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
-import java.io.OutputStream;
+import java.io.PipedInputStream;
+import java.io.PipedOutputStream;
 import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
-import java.nio.file.Files;
-import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
@@ -342,38 +339,37 @@ public interface ContainerState {
      * @param transferable file which is copied into the container
      * @param containerPath destination path inside the container
      */
-    @SneakyThrows(IOException.class)
+    @SneakyThrows({ IOException.class, InterruptedException.class })
     default void copyFileToContainer(Transferable transferable, String containerPath) {
         if (getContainerId() == null) {
             throw new IllegalStateException("copyFileToContainer can only be used with created / running container");
         }
 
-        Path tempFile = Files.createTempFile("tc-", "-copy");
-
         try (
-            OutputStream os = Files.newOutputStream(tempFile);
-            BufferedOutputStream bos = new BufferedOutputStream(os);
-            TarArchiveOutputStream tarArchive = new TarArchiveOutputStream(bos)
+            PipedOutputStream pipedOutputStream = new PipedOutputStream();
+            PipedInputStream pipedInputStream = new PipedInputStream(pipedOutputStream);
+            TarArchiveOutputStream tarArchive = new TarArchiveOutputStream(pipedOutputStream)
         ) {
-            tarArchive.setLongFileMode(TarArchiveOutputStream.LONGFILE_POSIX);
-            tarArchive.setBigNumberMode(TarArchiveOutputStream.BIGNUMBER_POSIX);
+            Thread thread = new Thread(() -> {
+                try {
+                    tarArchive.setLongFileMode(TarArchiveOutputStream.LONGFILE_POSIX);
+                    tarArchive.setBigNumberMode(TarArchiveOutputStream.BIGNUMBER_POSIX);
 
-            transferable.transferTo(tarArchive, containerPath);
-            tarArchive.finish();
-            bos.flush();
+                    transferable.transferTo(tarArchive, containerPath);
+                } finally {
+                    IOUtils.closeQuietly(tarArchive);
+                }
+            });
 
-            try (
-                InputStream is = Files.newInputStream(tempFile);
-                BufferedInputStream bis = new BufferedInputStream(is)
-            ) {
-                getDockerClient()
-                    .copyArchiveToContainerCmd(getContainerId())
-                    .withTarInputStream(bis)
-                    .withRemotePath("/")
-                    .exec();
-            }
-        } finally {
-            Files.deleteIfExists(tempFile);
+            thread.start();
+
+            getDockerClient()
+                .copyArchiveToContainerCmd(getContainerId())
+                .withTarInputStream(pipedInputStream)
+                .withRemotePath("/")
+                .exec();
+
+            thread.join();
         }
     }
 

--- a/core/src/main/java/org/testcontainers/containers/ContainerState.java
+++ b/core/src/main/java/org/testcontainers/containers/ContainerState.java
@@ -22,14 +22,17 @@ import org.testcontainers.utility.LogUtils;
 import org.testcontainers.utility.MountableFile;
 import org.testcontainers.utility.ThrowingFunction;
 
-import java.io.ByteArrayInputStream;
-import java.io.ByteArrayOutputStream;
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
 import java.io.File;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
+import java.io.OutputStream;
 import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
@@ -345,21 +348,32 @@ public interface ContainerState {
             throw new IllegalStateException("copyFileToContainer can only be used with created / running container");
         }
 
+        Path tempFile = Files.createTempFile("tc-", "-copy");
+
         try (
-            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
-            TarArchiveOutputStream tarArchive = new TarArchiveOutputStream(byteArrayOutputStream)
+            OutputStream os = Files.newOutputStream(tempFile);
+            BufferedOutputStream bos = new BufferedOutputStream(os);
+            TarArchiveOutputStream tarArchive = new TarArchiveOutputStream(bos)
         ) {
             tarArchive.setLongFileMode(TarArchiveOutputStream.LONGFILE_POSIX);
             tarArchive.setBigNumberMode(TarArchiveOutputStream.BIGNUMBER_POSIX);
 
             transferable.transferTo(tarArchive, containerPath);
             tarArchive.finish();
+            bos.flush();
 
-            getDockerClient()
-                .copyArchiveToContainerCmd(getContainerId())
-                .withTarInputStream(new ByteArrayInputStream(byteArrayOutputStream.toByteArray()))
-                .withRemotePath("/")
-                .exec();
+            try (
+                InputStream is = Files.newInputStream(tempFile);
+                BufferedInputStream bis = new BufferedInputStream(is)
+            ) {
+                getDockerClient()
+                    .copyArchiveToContainerCmd(getContainerId())
+                    .withTarInputStream(bis)
+                    .withRemotePath("/")
+                    .exec();
+            }
+        } finally {
+            Files.deleteIfExists(tempFile);
         }
     }
 

--- a/core/src/test/java/org/testcontainers/junit/FileOperationsTest.java
+++ b/core/src/test/java/org/testcontainers/junit/FileOperationsTest.java
@@ -3,15 +3,19 @@ package org.testcontainers.junit;
 import com.github.dockerjava.api.exception.NotFoundException;
 import org.apache.commons.io.FileUtils;
 import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.output.CountingOutputStream;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.TemporaryFolder;
 import org.testcontainers.TestImages;
+import org.testcontainers.containers.Container;
 import org.testcontainers.containers.GenericContainer;
 import org.testcontainers.containers.startupcheck.OneShotStartupCheckStrategy;
 import org.testcontainers.utility.MountableFile;
 
+import java.io.BufferedOutputStream;
 import java.io.File;
+import java.io.FileOutputStream;
 import java.io.IOException;
 
 import static org.assertj.core.api.Assertions.assertThat;
@@ -37,6 +41,43 @@ public class FileOperationsTest {
 
             File expectedFile = new File(mountableFile.getResolvedPath());
             assertThat(FileUtils.contentEquals(expectedFile, actualFile)).as("Files aren't same ").isTrue();
+        }
+    }
+
+    @Test
+    public void copyLargeFilesToContainer() throws Exception {
+        File tempFile = temporaryFolder.newFile();
+        try (
+            GenericContainer alpineCopyToContainer = new GenericContainer(TestImages.ALPINE_IMAGE) //
+                .withCommand("sleep", "infinity")
+        ) {
+            alpineCopyToContainer.start();
+            final long byteCount;
+            try (
+                FileOutputStream fos = new FileOutputStream(tempFile);
+                CountingOutputStream cos = new CountingOutputStream(fos);
+                BufferedOutputStream bos = new BufferedOutputStream(cos)
+            ) {
+                for (int i = 0; i < 0x4000; i++) {
+                    byte[] bytes = new byte[0xFFFF];
+                    bos.write(bytes);
+                }
+                bos.flush();
+                byteCount = cos.getByteCount();
+            }
+            final MountableFile mountableFile = MountableFile.forHostPath(tempFile.getPath());
+            final String containerPath = "/test.bin";
+            alpineCopyToContainer.copyFileToContainer(mountableFile, containerPath);
+
+            final Container.ExecResult execResult = alpineCopyToContainer.execInContainer( //
+                "stat",
+                "-c",
+                "%s",
+                containerPath
+            );
+            assertThat(execResult.getStdout()).isEqualToIgnoringNewLines(Long.toString(byteCount));
+        } finally {
+            tempFile.delete();
         }
     }
 

--- a/core/src/test/java/org/testcontainers/junit/FileOperationsTest.java
+++ b/core/src/test/java/org/testcontainers/junit/FileOperationsTest.java
@@ -48,7 +48,7 @@ public class FileOperationsTest {
     public void copyLargeFilesToContainer() throws Exception {
         File tempFile = temporaryFolder.newFile();
         try (
-            GenericContainer alpineCopyToContainer = new GenericContainer(TestImages.ALPINE_IMAGE) //
+            GenericContainer<?> alpineCopyToContainer = new GenericContainer<>(TestImages.ALPINE_IMAGE) //
                 .withCommand("sleep", "infinity")
         ) {
             alpineCopyToContainer.start();

